beluga_ros/include/beluga_ros/amcl.hpp

```diff
@@ -0,0 +1,212 @@
+    }
+
+    if (!update_policy_(base_pose_in_odom) && !force_update_) {
+      return std::nullopt;
+    }
+
+    // TODO(nahuel): Remove this once we update the measurement type.
+    auto measurement = laser_scan.points_in_cartesian_coordinates() |  //
+                       ranges::views::transform([&laser_scan](const auto& p) {
+                         const auto result = laser_scan.origin() * Sophus::Vector3d{p.x(), p.y(), 0};
+                         return std::make_pair(result.x(), result.y());
+                       }) |
+                       ranges::to<std::vector>;
+
+    std::visit(
+        [&, this](auto& policy, auto& motion_model, auto& sensor_model) {
+          particles_ |=
+              beluga::actions::propagate(policy, motion_model(control_action_window_ << base_pose_in_odom)) |  //
+              beluga::actions::reweight(policy, sensor_model(std::move(measurement))) |                        //
+              beluga::actions::normalize(policy);
+        },
+        execution_policy_, motion_model_, sensor_model_);
+
+    random_state_probability_ = random_probability_estimator_(particles_);
+
+    if (resample_policy_(particles_)) {
+      auto random_state = ranges::compose(beluga::make_from_state<particle_type>, std::ref(map_distribution_));
+
+      if (random_state_probability_ > 0.0) {
+        random_probability_estimator_.reset();
+      }
+
+      particles_ |= beluga::views::sample |
+                    beluga::views::random_intersperse(std::move(random_state), random_state_probability_) |
+                    beluga::views::take_while_kld(
+                        spatial_hasher_,        //
+                        params_.min_particles,  //
+                        params_.max_particles,  //
+                        params_.kld_epsilon,    //
+                        params_.kld_z) |
+                    beluga::actions::assign;
+    }
+
+    force_update_ = false;
+    return beluga::estimate(beluga::views::states(particles_), beluga::views::weights(particles_));
+  }
+
+  /// Force a manual update of the particles on the next iteration of the filter.
+  void force_update() { force_update_ = true; }
+
+ private:
+  beluga::TupleVector<particle_type> particles_;
+
+  AmclParams params_;
+  beluga::MultivariateUniformDistribution<Sophus::SE2d, beluga_ros::OccupancyGrid> map_distribution_;
+  motion_model_variant motion_model_;
+  sensor_model_variant sensor_model_;
+  execution_policy_variant execution_policy_;
+
+  beluga::spatial_hash<Sophus::SE2d> spatial_hasher_;
+  beluga::ThrunRecoveryProbabilityEstimator random_probability_estimator_;
+  beluga::any_policy<Sophus::SE2d> update_policy_;
+  beluga::any_policy<decltype(particles_)> resample_policy_;
+
+  beluga::RollingWindow<Sophus::SE2d, 2> control_action_window_;
+
+  bool force_update_{true};
+
+  std::optional<std::pair<Sophus::SE2d, Eigen::Matrix3d>> latest_estimate_;
+  std::optional<Sophus::SE2d> latest_odom_transform_in_map_;
```

Comment:
@nahueespinosa unused?

Reply:
Yes! Sorry, fixed in https://github.com/Ekumen-OS/beluga/pull/327/commits/af14ee895f234d1b2b21c1a5307a019b21edcf97.